collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java

```diff
@@ -0,0 +1,139 @@
+import java.io.OutputStream;
+import java.util.Arrays;
+
+import timber.log.Timber;
+
+import static junit.framework.Assert.assertEquals;
+
+/**
+ * This test has been created in order to check indices while navigating through a form.
+ * It's especially important while navigate through a form which contains nested groups and if we
+ * use groups with field-list appearance because in that case we need to collect all indices of
+ * questions we want to display on one page (we need to recursively get all indices contained in
+ * such a group and its children). It might be also tricky when navigate backwards because then we
+ * need to navigate to an index of the first question of all we want to display on one page.
+ */
+@RunWith(Parameterized.class)
+public class FormNavigationTestCase {
+
+    private static final String FORMS_DIRECTORY = "/odk/forms/";
+
+    @Parameters(name = "{0}")
+    public static Iterable<Object[]> data() {
+        return Arrays.asList(new Object[][] {
+                {"simpleFieldList.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListInFieldList.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"regularGroupWithFieldListGroupInside.xml", new String[] {"-1, ", "0, 0, ", "-1, ", "0, 0, "}},
+                {"twoNestedRegularGroups.xml", new String[] {"-1, ", "0, 0, 0, ", "0, 0, 1, ", "0, 0, 2, ", "-1, ", "0, 0, 2, "}},
+                {"regularGroupWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 1, 1, ", "-1, ", "0, 1, 1, "}},
+                {"regularGroupWithQuestionsAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "}},
+                {"fieldListWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListWithQuestionsAndRegularGroupsInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"threeNestedFieldListGroups.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}}
+        });
+    }
+
+    private String formName;
+    private String[] expectedIndices;
+
+    public FormNavigationTestCase(String formName, String[] expectedIndices) {
+        this.formName = formName;
+        this.expectedIndices = expectedIndices;
+    }
+
+    @Test
+    public void formNavigationTestCase() {
+        testIndices(formName, expectedIndices);
+    }
+
+    private void testIndices(String formName, String[] expectedIndices) {
+        try {
+            prepareFile(formName);
+        } catch (IOException e) {
+            Timber.i(e);
+        }
+        FormLoaderTask formLoaderTask = new FormLoaderTask(formPath(formName), null, null);
+        formLoaderTask.setFormLoaderListener(new FormLoaderListener() {
+            @Override
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    for (int i = 0; i < expectedIndices.length - 1; i++) {
```

Comment:
I’m finding it challenging to understand the “- 1”, “- 2”, “stepToNext...”, “stepToPrevious...”, “+ 1”, and why there are two assertEquals(expectedIndices...) each time through the loop. I wonder if a few comments in here would help me and others.

Reply:
We have two asserts because I check the current index and the index after navigating (back or forward). I added a comment.

Reply:
I've just added comments.